src/io/delimited.cr

```diff
@@ -44,8 +44,182 @@ class IO::Delimited < IO
+        return read_from_active_delimited_buffer(slice)
+      end
+    end
 
+    first_byte = @read_delimiter[0]
+
+    # If we have something in the active delimiter buffer
+    unless @active_delimiter_buffer.empty?
+      # This is the rest of the delimiter we have to match
+      delimiter_remaining = @read_delimiter[@active_delimiter_buffer.size..]
+
+      # This is how much we can actually match of that (peek might not have enough data!)
+      min_size = Math.min(delimiter_remaining.size, peek.size)
+
+      # See if what remains to match in the delimiter matches whatever
+      # we have in peek, limited to what's available
+      if delimiter_remaining[0, min_size] == peek[0, min_size]
+        # If peek has enough data to match the entire rest of the delimiter...
+        if peek.size >= delimiter_remaining.size
+          # We found the delimiter!
+          @io.skip(min_size)
+          @active_delimiter_buffer = Bytes.empty
+          @finished = true
+          return 0
+        else
+          # Copy the remaining of peek to the active delimiter buffer for now
+          (@delimiter_buffer + @active_delimiter_buffer.size).copy_from(peek)
+          @active_delimiter_buffer = @delimiter_buffer[0, @active_delimiter_buffer.size + peek.size]
+
+          # Skip whatever we had in peek, and try reading more
+          @io.skip(peek.size)
+          return read_internal(slice)
+        end
+      else
+        # No match.
+        # We first need to check if the delimiter could actually start in this active buffer.
+        next_index = @active_delimiter_buffer.index(first_byte, 1)
+
+        # We read up to that new match, if any, or the entire buffer
+        read_bytes = next_index || @active_delimiter_buffer.size
+
+        slice.copy_from(@active_delimiter_buffer[0, read_bytes])
+        slice += read_bytes
+        @active_delimiter_buffer += read_bytes
+        return read_bytes + read_internal(slice)
+      end
+    end
+
+    index =
+      if slice.size == 1
+        # For a size of 1, this is much faster
+        first_byte == peek[0] ? 0 : nil
+      elsif slice.size < peek.size
+        peek[0, slice.size].index(first_byte)
+      else
+        peek.index(first_byte)
+      end
+
+    # If we can't find the delimiter's first byte we can just read from peek
+    unless index
+      # If we have more in peek than what we need to read, read all of that
+      if peek.size >= slice.size
+        if slice.size == 1
+          # For a size of 1, this is much faster
+          slice[0] = peek[0]
+        else
+          slice.copy_from(peek[0, slice.size])
+        end
+        @io.skip(slice.size)
+        return slice.size
+      else
+        # Otherwise, read from peek for now
+        slice.copy_from(peek)
+        @io.skip(peek.size)
+        return peek.size
+      end
+    end
+
+    # If the matching byte is past the data we need to read,
+    # we can fill the entire slice
+    if index >= slice.size
```

Comment:
Is this condition even possible? The code for finding the index seems to search only in the range < slice.size

Reply:
Looking at:

```crystal
    index =
      if slice.size == 1
        # For a size of 1, this is much faster
        first_byte == peek[0] ? 0 : nil
      elsif slice.size < peek.size
        peek[0, slice.size].index(first_byte)
      else
        peek.index(first_byte)
      end
```

it seems `index` can be anywhere in `0...peek.size`, right? And peek.size might be greater than slice.size. That said, commenting this still works, so I might be missing a spec. What would be a spec for this scenario?

Reply:
If `peek.size > slice.size`, the second branch is chosen, which limits the search range to `peek[0, slice.size]`: Thus the index can't be bigger than `slice.size`.

Reply:
You are totally right! Fixed.